src/app/fdctl/run/tiles/fd_poh.c

```diff
@@ -842,36 +847,36 @@ fd_ext_poh_reached_leader_slot( ulong * out_leader_slot,
     return 1;
   }
 
-  if( FD_LIKELY( ctx->next_leader_slot>=1UL ) ) {
-    fd_epoch_leaders_t * leaders = fd_stake_ci_get_lsched_for_slot( ctx->stake_ci, ctx->next_leader_slot-1UL ); /* Safe to call from Rust */
-    if( FD_LIKELY( leaders ) ) {
-      fd_pubkey_t const * leader = fd_epoch_leaders_get( leaders, ctx->next_leader_slot-1UL ); /* Safe to call from Rust */
-      if( FD_LIKELY( leader ) ) {
-        if( FD_UNLIKELY( !memcmp( leader->uc, ctx->identity_key.uc, 32UL ) ) ) {
-          /* We were the leader in the previous slot, so also no need for
-             a grace period.  We wouldn't get here if we were still
-             processing the prior slot so begin new one immediately. */
-          fd_ext_poh_write_unlock();
-          return 1;
-        }
+  long now_ns = fd_log_wallclock();
+  long expected_start_time_ns = ctx->reset_slot_start_ns + (long)((double)(ctx->next_leader_slot-ctx->reset_slot)*ctx->slot_duration_ns);
+
+  if( FD_UNLIKELY( now_ns<expected_start_time_ns+1200000000L ) ) {
+    int wait_graceful = 0;
+    for( ulong i=0UL; i<ctx->active_descendants_len; i++ ) {
+      if( FD_LIKELY( ctx->active_descendants[ i ]>=ctx->reset_slot && ctx->active_descendants[ i ]<ctx->next_leader_slot ) ) {
+        wait_graceful = 1;
+      } else if( FD_UNLIKELY( ctx->active_descendants[ i ]>=ctx->next_leader_slot ) ) {
+        /* Oops, we waited too long and a leader after us started
+           publishing to try and skip us.  Just start our slot
+           immediately, we might win ... */
+        wait_graceful = 0;
```

Comment:
I think we can do all this logic only knowing max(active_descendants), right?. That would make `fd_ext_poh_update_active_descendants` a lot simpler and faster